src/expr/dtype.cpp

```diff
@@ -557,6 +565,133 @@ Node DType::mkGroundTermInternal(TypeNode t, bool isValue) const
   return groundTerm;
 }
 
+void DType::getStrictSubfieldTypes(
+    std::unordered_set<TypeNode, TypeNodeHashFunction>& types,
+    std::map<TypeNode, bool>& processed,
+    bool isStrictC) const
+{
+  std::map<TypeNode, bool>::iterator it = processed.find(d_self);
+  if (it != processed.end())
+  {
+    if (it->second || (!isStrictC && !it->second))
+    {
+      // already processed as a strict subfield type, or already processed
+      // as a non-strict subfield type and isStrictC is false.
+      return;
+    }
+  }
+  processed[d_self] = isStrictC;
+  for (std::shared_ptr<DTypeConstructor> ctor : d_constructors)
+  {
+    for (unsigned j = 0, nargs = ctor->getNumArgs(); j < nargs; ++j)
+    {
+      TypeNode tn = ctor->getArgType(j);
+      if (tn.isDatatype())
+      {
+        // special case for datatypes, we must recurse to collect subfield types
+        if (!isStrictC)
+        {
+          // since we aren't adding it to types below, we add its strict
+          // subfield types here.
+          const DType& dt = tn.getDType();
+          dt.getStrictSubfieldTypes(types, processed, false);
+        }
+        if (tn.isParametricDatatype() && !isStrictC)
+        {
+          // (instantiated) parametric datatypes have an AST structure:
+          //  (PARAMETRIC_DATATYPE D T1 ... Tn)
+          // where D is the uninstantiated datatype type.  We should not view D
+          // as a strict subfield type of tn. Thus, we need a special case here
+          // which ignores the first child, when isStrictC is false.
+          for (unsigned i = 1, nchild = tn.getNumChildren(); i < nchild; i++)
+          {
+            expr::getComponentTypes(tn[i], types);
+          }
+          continue;
+        }
+      }
+      bool hasTn = types.find(tn) != types.end();
+      Trace("datatypes-init")
+          << "Collect subfield types " << tn << ", hasTn=" << hasTn
+          << ", isStrictC=" << isStrictC << std::endl;
+      expr::getComponentTypes(tn, types);
+      if (!isStrictC && !hasTn)
+      {
+        // the top-level type is not a strict subfield type
```

Comment:
Extend comment: the top-level type is added by getComponentTypes, so remove it if it was not already listed in types